### PR TITLE
Implement recv() and recvmsg()

### PR DESCRIPTION
**Pull request by DK-DARKmatter**
Implement recv() and recvmsg() function. But tls part is not implemented
and needs to be done in the future.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -263,8 +272,16 @@ pub extern "C" fn recvfrom(
...
-}
+    if addr.is_null() && addr_len.is_null() {
+        return recv(fd, buf, n, flags);
+    }
 
+    return match INDEX.get(fd) {
+        None => next!(recvfrom(fd, buf, n, flags, addr, addr_len)),
+        Some(s) => -1,
+    };
+}
```
Comment (npmccallum) [inline src/lib.rs:282]: This code does not match the behavior of the original function: https://github.com/enarx/tlssock/blob/master/lib/tlssock.c#L234

**Review by npmccallum — CHANGES_REQUESTED** [review]

**Comment by DK-DARKmatter** [comment]
Does this look better now?

File: src/lib.rs [hunk 2]
```diff
@@ -263,8 +272,19 @@ pub extern "C" fn recvfrom(
...
 ) -> isize {
-}
+    if addr.is_null() && addr_len.is_null() {
+        return recv(fd, buf, n, flags);
+    }
 
+    return match INDEX.get(fd) {
+        None => next!(recvfrom(fd, buf, n, flags, addr, addr_len)),
+        Some(s) => {
+            errno::set_errno(libc::ENOSYS);
```
Comment (npmccallum) [inline src/lib.rs]: Use the `error()` function.

**Review by npmccallum — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 3]
```diff
@@ -243,18 +243,27 @@ pub extern "C" fn connect(
...
+
+    return match flags {
+        0 => 0,//TODO: tls_read() implement
+        _ => error(libc::EINVAL, -1),
+    };
+}
 
 
 #[no_mangle]
+use libc::{sockaddr, socklen_t};
```
Comment (npmccallum) [inline src/lib.rs]: Don't import the types here. Use `libc::sockaddr` instead below.

**Review by npmccallum — CHANGES_REQUESTED** [review]

**Review by npmccallum — APPROVED** [review]

**Pull request merged**